pkg/autoscaler/stats_scraper.go

```diff
@@ -52,48 +60,119 @@ var cacheDisabledClient = &http.Client{
+	}
+	if dynamicConfig == nil {
+		return nil, errors.New("empty point of DynamicConfig")
+	}
+	if endpointsInformer == nil {
+		return nil, errors.New("empty interface of EndpointsInformer")
+	}
+	if httpClient == nil {
+		return nil, errors.New("empty point of HTTP client")
+	}
 	revName := metric.Labels[serving.RevisionLabelKey]
 	if revName == "" {
 		return nil, fmt.Errorf("no Revision label found for Metric %s", metric.Name)
 	}
 
 	serviceName := reconciler.GetServingK8SServiceNameForObj(revName)
 	return &ServiceScraper{
-		httpClient: httpClient,
-		url:        fmt.Sprintf("http://%s.%s:%d/metrics", serviceName, metric.Namespace, v1alpha1.RequestQueueMetricsPort),
-		metricKey:  NewMetricKey(metric.Namespace, metric.Name),
-		logger:     logger,
+		httpClient:      httpClient,
+		endpointsLister: endpointsInformer.Lister(),
+		url:             fmt.Sprintf("http://%s.%s:%d/metrics", serviceName, metric.Namespace, v1alpha1.RequestQueueMetricsPort),
+		metricKey:       NewMetricKey(metric.Namespace, metric.Name),
+		namespace:       metric.Namespace,
+		revisionService: reconciler.GetServingK8SServiceNameForObj(revName),
+		cachedReadyPods: readyPods{},
 	}, nil
 }
 
 // Scrape call the destination service then send it
 // to the given stats chanel
-func (s *ServiceScraper) Scrape(statsCh chan<- *StatMessage) {
+func (s *ServiceScraper) Scrape(ctx context.Context, statsCh chan<- *StatMessage) {
+	logger := logging.FromContext(ctx)
+
+	readyPodsCount, err := s.readyPodsCount(time.Now())
+	if err != nil {
+		logger.Errorw("Failed to get Endpoints via K8S Lister", zap.Error(err))
+		return
+	}
+
+	if readyPodsCount == 0 {
+		logger.Debug("No ready pods found, not to scrape.")
+		return
+	}
+
 	stat, err := s.scrapeViaURL()
 	if err != nil {
-		s.logger.Errorw("Failed to get metrics", zap.Error(err))
+		logger.Errorw("Failed to get metrics", zap.Error(err))
 		return
 	}
 
-	s.sendStatMessage(*stat, statsCh)
+	// Assume traffic is route to pods evenly. A particular pod can stand for
+	// other pods, i.e. other pods have similar concurrency and QPS.
+	// Hide the actual pods behind scraper and send only one stat for all the
+	// customer pods per scraping. The pod name is set to a unique value, i.e.
+	// scraperPodName so in autoscaler all stats are either from activator or
+	// scraper.
+	newStat := Stat{
+		Time:                      stat.Time,
+		PodName:                   scraperPodName,
+		AverageConcurrentRequests: stat.AverageConcurrentRequests * float64(readyPodsCount),
+		RequestCount:              stat.RequestCount * int32(readyPodsCount),
+	}
+
+	s.sendStatMessage(newStat, statsCh)
+}
+
+// readyPodsCount returns the ready IP count in the K8S Endpoints object for a Revision
+// via K8S Informer. This is same as ready Pod count.
+func (s *ServiceScraper) readyPodsCount(now time.Time) (int, error) {
+	if s.cachedReadyPods.time.Add(cacheTimeout).After(now) {
+		return s.cachedReadyPods.count, nil
+	}
```

Comment:
I get why you're using caching here. Is it necessary though? Shouldn't the informer you pass in already do caching internally?

Reply:
Talked to @vagababov, the performance gain for caching in memory ourselves is probably negligible. Removed.